listing/listing.go

```diff
@@ -0,0 +1,195 @@
+// page of items, returning the deserialized response and error. getItems
+// selects the items being iterated over from the response. getNextReq is used
+// to get the next request to be used in the next page. If the returned value
+// is nil, then there is no next page to fetch.
+func NewIterator[Req, Resp, T any](
+	nextReq *Req,
+	getNextPage func(context.Context, Req) (Resp, error),
+	getItems func(Resp) []T,
+	getNextReq func(Resp) *Req,
+) *iteratorImpl[Req, Resp, T] {
+	return &iteratorImpl[Req, Resp, T]{
+		nextReq:     nextReq,
+		getNextPage: getNextPage,
+		getItems:    getItems,
+		getNextReq:  getNextReq,
+	}
+}
+
+func (i *iteratorImpl[Req, Resp, T]) loadNextPageIfNeeded(ctx context.Context) error {
+	if i.currentPageIdx < len(i.currentPage) {
+		return nil
+	}
+	if i.nextReq == nil {
+		i.currentPage = nil
+		i.currentPageIdx = 0
+		return nil
+	}
+	if i.lastErr != nil {
+		return i.lastErr
+	}
+
+	// Keep loading pages while we have a next request and the current page is
+	// empty.
+	i.currentPage = nil
+	for i.nextReq != nil && len(i.currentPage) == 0 {
+		resp, err := i.getNextPage(ctx, *i.nextReq)
+		i.lastErr = err
+		if err != nil {
+			return err
+		}
+		items := i.getItems(resp)
+		i.currentPage = items
+		i.currentPageIdx = 0
+		i.nextReq = i.getNextReq(resp)
+	}
+	return nil
+}
+
+func (i *iteratorImpl[Req, Resp, T]) Next(ctx context.Context) (T, error) {
+	var t T
+	err := i.loadNextPageIfNeeded(ctx)
+	if err != nil {
+		return t, err
+	}
+	if i.currentPageIdx >= len(i.currentPage) {
+		return t, ErrNoMoreItems
+	}
+	item := i.currentPage[i.currentPageIdx]
+	i.currentPageIdx++
+	return item, nil
+}
+
+func (i *iteratorImpl[Req, Resp, T]) HasNext(ctx context.Context) bool {
+	err := i.loadNextPageIfNeeded(ctx)
+	if err != nil {
+		return true
+	}
+	return i.currentPageIdx < len(i.currentPage)
+}
+
+type dedupeIteratorImpl[T any, Id comparable] struct {
+	it      Iterator[T]
+	getId   func(T) Id
+	seen    map[Id]struct{}
+	current *T
+}
+
+func NewDedupeIterator[T any, Id comparable](it Iterator[T], getId func(T) Id) *dedupeIteratorImpl[T, Id] {
+	return &dedupeIteratorImpl[T, Id]{
+		it:    it,
+		getId: getId,
+		seen:  make(map[Id]struct{}),
+	}
+}
+
+func (i *dedupeIteratorImpl[T, Id]) Next(ctx context.Context) (T, error) {
+	if i.current != nil {
+		t := *i.current
+		i.current = nil
+		return t, nil
+	}
+	for {
+		t, err := i.it.Next(ctx)
+		if err != nil {
+			return t, err
+		}
+		id := i.getId(t)
+		if _, ok := i.seen[id]; !ok {
+			i.seen[id] = struct{}{}
+			return t, nil
+		}
+	}
+}
+
+func (i *dedupeIteratorImpl[T, Id]) HasNext(ctx context.Context) bool {
+	if i.current != nil {
+		return true
+	}
+	for {
+		t, err := i.it.Next(ctx)
+		if errors.Is(err, ErrNoMoreItems) {
+			return false
+		}
+		if err != nil {
+			return true
+		}
+		id := i.getId(t)
+		if _, ok := i.seen[id]; !ok {
+			i.current = &t
+			i.seen[id] = struct{}{}
+			return true
+		}
+	}
```

Comment:
Can you include some comments for this one? From our conversation I now know why we need to keep this element around, but it would be great if a future reader can refer to a short description vs decoding it themselves.